src/persist-client/src/examples/persistent_source_example.rs

```diff
@@ -0,0 +1,1395 @@
+    // internally in the persist client.
+    // let (blob, consensus) = location
+    //     .open_unreliable(Instant::now() + NO_TIMEOUT)
+    //     .await?;
+    PersistClient::new(blob, consensus).await
+}
+
+mod api {
+    use std::fmt::Debug;
+
+    use async_trait::async_trait;
+    use mz_persist::location::ExternalError;
+    use timely::progress::{Antichain, ChangeBatch};
+
+    #[derive(Debug)]
+    pub struct Message<T, P> {
+        pub capability: ChangeBatch<T>,
+        pub payload: P,
+    }
+
+    #[derive(Debug)]
+    pub struct Batch<D, T> {
+        pub updates: Vec<(D, T)>,
+
+        /// The lowest timestamp(s) of updates in this batch.
+        pub lower: Antichain<T>,
+
+        /// The highest timestamp(s) of updates in this batch. This is different from a
+        /// _differential upper_, which is exclusive and denotes the frontier at which new updates
+        /// would happen. For reclocking puposes (finding/minting a matching timestamp binding) we
+        /// are interested in the largest source timestamps (think Kafka partition/offset pairs)
+        /// for which we need to find a binding.
+        pub inclusive_upper: Antichain<T>,
+    }
+
+    /// `TT` is `ToTime`.
+    ///
+    /// We carry information about the source frontier inline with batches, so that we can know the
+    /// frontier in the task that does emission.
+    #[derive(Debug)]
+    pub struct TimestampedBatch<D, TT> {
+        pub updates: Vec<(D, TT)>,
+
+        pub inclusive_upper: Antichain<TT>,
+    }
+
+    #[async_trait]
+    pub trait Source: Clone {
+        type SourceTimestamp: timely::PartialOrder + PartialOrd + Debug;
+        type D: Debug;
+
+        /// Reads up to `max_num_records` updates from the source. Returns `None` if no data is
+        /// available.
+        ///
+        /// We could think about returning an empty batch with the same lower/inclusive_upper as
+        /// last time. Or we could think about this method just blocking forever, but it doesn't
+        /// feel like the correct thing to do. But, idk... ¯\_(ツ)_/¯
+        async fn read_batch(
+            &mut self,
+            max_num_records: usize,
+        ) -> Option<Batch<Self::D, Self::SourceTimestamp>>;
+
+        /// Returns the current frontier of this source. All future updates will be at times that
+        /// are beyond this frontier.
+        async fn frontier(&self) -> Antichain<Self::SourceTimestamp>;
+    }
+
+    #[async_trait]
+    pub trait Timestamper {
+        type FromTimestamp: Debug + timely::PartialOrder;
+        type ToTimestamp: Debug + Copy;
+
+        async fn get_bindings(
+            &mut self,
+            lower: Antichain<Self::FromTimestamp>,
+            upper: Antichain<Self::FromTimestamp>,
+        ) -> TimestampBindings<Self::FromTimestamp, Self::ToTimestamp>;
+
+        /// All bindings that this `Timestamper` knows about. It might be that there are more
```

Comment:
nit: is this all the bindings that `Timestamper` knows _are durable_?

Reply:
yes